lld/COFF/Writer.cpp

```diff
@@ -1668,21 +1680,38 @@ template <typename PEHeaderTy> void Writer::writeHeader() {
   // When run under Windows, the loader looks at AddressOfNewExeHeader and uses
   // the PE header instead.
   Configuration *config = &ctx.config;
+
   uint8_t *buf = buffer->getBufferStart();
   auto *dos = reinterpret_cast<dos_header *>(buf);
-  buf += sizeof(dos_header);
-  dos->Magic[0] = 'M';
-  dos->Magic[1] = 'Z';
-  dos->UsedBytesInTheLastPage = dosStubSize % 512;
-  dos->FileSizeInPages = divideCeil(dosStubSize, 512);
-  dos->HeaderSizeInParagraphs = sizeof(dos_header) / 16;
-
-  dos->AddressOfRelocationTable = sizeof(dos_header);
-  dos->AddressOfNewExeHeader = dosStubSize;
 
   // Write DOS program.
-  memcpy(buf, dosProgram, sizeof(dosProgram));
-  buf += sizeof(dosProgram);
+  if (config->dosStub) {
+    memcpy(buf, config->dosStub->getBufferStart(),
+           config->dosStub->getBufferSize());
+    // MS link.exe accepts an invalid `e_lfanew` (AddressOfNewExeHeader) and
+    // updates it automatically. Replicate the same behaviour.
+    dos->AddressOfNewExeHeader = config->dosStub->getBufferSize();
+    buf += config->dosStub->getBufferSize();
+    // Unlike MS link.exe, LLD accepts non-8-byte-aligned stubs.
+    // In that case, we add zero paddings ourselves.
+    buf += (8 - (config->dosStub->getBufferSize() % 8)) % 8;
```

Comment:
Better to use alignTo().

And how is buf's size computed? Can we be sure there's room for this potential extra padding?

Reply:
the final file size (`fileSize`) is computed in `Writer::assignAddresses()` based on `sizeofHeaders` and the caller `Writer::run()` checks buf size: `if (fileSize > UINT32_MAX)` before committing to a final file. So technically there must be an enough space for the paddings (at most it'd be 7 bytes long), except the case where the size of given stub input exceeds gigabytes (wow). Needless to say, such considered an expected behaviour imo.

Reply:
Sounds good, thanks for confirming.